### PR TITLE
test(trtllm): stabilize health-check priority coverage [DYN-4061]

### PR DESCRIPTION
**Pull request by knarangN**
## Summary

Fixes DYN-4061 by keeping the startup-sensitive H100 aggregated health-check test in the existing sequential GPU stage.

## Root cause

In the attached failing Nightly run (`32231278613`, job `96005427436`), the test started on GPU1 with seven other subprocesses. Its worker reached TRT-LLM model registration and `MpiPoolSession` startup, but never registered an endpoint; the worker `/health` endpoint returned 503 until pytest hit the unchanged 300-second timeout. No canary was dispatched.

The failure is concurrent TRT-LLM MPI engine startup contention in the VRAM-parallel stage, before health-check dispatch or priority handling. The priority behavior itself already exists: `TrtllmHealthCheckPayload` sets `priority=1.0`, `HandlerBase` forwards it to `generate_async`, and `test_health_check_gets_priority_1` directly asserts that contract.

## Changes

- Comment out `profiled_vram_gib(3.9)` for this test so the existing workflow excludes it from the GPU-parallel stage and runs it in the sequential stage.
- Document the startup-contention reason next to the marker.
- Keep the test scenario, request, production code, scheduler, and 300-second timeout unchanged.
- Add no new marker or scheduling mechanism.

## Validation

- Repository pre-commit checks for the final changed file: passed, including the pytest marker report when run with host networking.
- Isort, Black, Flake8, Ruff, codespell, file hygiene, and `git diff --check`: passed.
- DCO: author, committer, and `Signed-off-by` are `Kavita Narang <knarang@nvidia.com>`.
- The temporary pre-merge promotion previously passed the target test in 167 seconds; exact-head H100 Nightly validation is still required for the final commit.

## Related issue

- Internal tracking: DYN-4061

### CODE REVIEW

File: tests/utils/pytest_parallel_gpu.py [hunk 1]
```diff
@@ -389,10 +405,59 @@ def _cap(gi: int) -> float:
...
+                and gi not in reserved_req
+                and gi not in exclusive_reserved
+            ]
+            if drain_candidates:
+                gi = min(
+                    drain_candidates,
+                    key=lambda g: (tentative[g].count, tentative[g].budget),
+                )
+                exclusive_reserved.add(gi)
+            continue
```
[the remark this answers is not in the document]
Comment (knarangN) [reply tests/utils/pytest_parallel_gpu.py]: Addressed in c6e7be4c0b. Drain candidates now require scheduler-tracked VRAM work, so an externally occupied but scheduler-idle GPU is not falsely reserved as drainable. The runner also has a bounded 60-second no-progress guard that reports GPU usage and queued tests before returning an error, with unit coverage.

File: tests/serve/test_trtllm.py [hunk 2]
```diff
@@ -823,9 +823,15 @@ def test_chat_only_aggregated_with_test_logits_processor(
 @pytest.mark.gpu_1
 @pytest.mark.trtllm
 @pytest.mark.core
-@pytest.mark.nightly
+# TODO: revert to pytest.mark.nightly after pre_merge validation
+# on this PR (see .ai/ci-guidelines.md).
+@pytest.mark.pre_merge
```
[the remark this answers is not in the document]
Comment (knarangN) [reply tests/serve/test_trtllm.py]: Addressed in c6e7be4c0b. The temporary pre_merge promotion and TODO were removed, and test_aggregated_health_check_priority is restored to pytest.mark.nightly. The trusted promotion run passed this test in 167 seconds; a final-head branch nightly will validate the restored selector.

File: tests/utils/pytest_parallel_gpu.py [hunk 3]
```diff
@@ -389,10 +405,59 @@ def _cap(gi: int) -> float:
...
         if test.profiled_gib <= 0:
-            gi = min(gpu_states, key=lambda g: tentative[g].count)
+            candidates = [
+                gi
+                for gi in gpu_states
+                if not tentative[gi].exclusive and gi not in exclusive_reserved
+            ]
+            if not candidates:
+                continue
+            gi = min(candidates, key=lambda g: tentative[g].count)
```
[the remark this answers is not in the document]
Comment (knarangN) [reply tests/utils/pytest_parallel_gpu.py]: Addressed in c6e7be4c0b. Zero-VRAM fillers now bypass exclusive and drain reservations, and they are no longer counted as per-GPU VRAM processes, so an exclusive test can start while a filler is already running. Added regression coverage for running, draining, and already-running-filler cases; the focused scheduler suite passes 21/21.

File: tests/utils/pytest_parallel_gpu.py [hunk 4]
```diff
@@ -389,12 +423,54 @@ def _cap(gi: int) -> float:
...
+            candidates = []
+            for gi, gs in gpu_states.items():
+                ts = tentative[gi]
+                actual_used = gs.total_gib - ts.free
+                if (
+                    ts.count == 0
+                    and not ts.exclusive
+                    and gi not in reserved_req
+                    and gi not in exclusive_reserved
+                    and actual_used + test.profiled_gib <= gs.total_gib
```
Comment (nv-tusharma) [inline tests/utils/pytest_parallel_gpu.py]: [P1] Exclusive tests can launch on a non-idle GPU

`ts.count == 0` only proves that the scheduler has no tracked VRAM process on the card. The live-memory predicate still permits launch whenever the exclusive test fits; for example, an 80-GiB GPU with 40 GiB already allocated will accept this 3.9-GiB TRT-LLM test. An external process or an orphaned MPI engine left after a failed subprocess or retry can therefore recreate the concurrent startup contention this marker is intended to prevent.

Could we require live usage to be near a captured idle baseline, with tolerance, before launching an exclusive test? A regression case with `running_count=0` and `actual_free < total` would cover this path.

File: tests/serve/test_trtllm.py [hunk 5]
```diff
@@ -826,6 +826,10 @@ def test_chat_only_aggregated_with_test_logits_processor(
 @pytest.mark.nightly
 @pytest.mark.profiled_vram_gib(3.9)
 @pytest.mark.requested_trtllm_kv_tokens(2592)
+# Packed H100 runs can start several TRT-LLM MPI engines on one GPU at once;
+# reserve this GPU so the test measures canary readiness instead of startup
+# contention.
+@pytest.mark.gpu_parallel_exclusive
```
Comment (nv-tusharma) [inline tests/serve/test_trtllm.py]: instead of creating a seperate marker, can we just keep this keep this in the sequential stage and comment out the vram for now?

Comment (nv-tusharma) [reply tests/serve/test_trtllm.py]: Want to avoid creating unnecessary markers.

Comment (knarangN) [reply tests/serve/test_trtllm.py]: Addressed in `4cecc8646a`. I removed the new marker and scheduler changes, and commented out `profiled_vram_gib(3.9)` for this startup-sensitive test so it falls through to the existing sequential GPU stage.

File: tests/utils/pytest_parallel_gpu.py (hunk 4 above)
Comment (knarangN) [reply tests/utils/pytest_parallel_gpu.py]: Addressed in `4cecc8646a`. I removed `gpu_parallel_exclusive` and the custom exclusive scheduling path entirely. The health-check test is now intentionally unprofiled, so the existing workflow runs it in the sequential GPU stage and the non-idle exclusive-launch case no longer applies.

File: tests/serve/test_trtllm.py (hunk 5 above)
Comment (knarangN) [reply tests/serve/test_trtllm.py]: Follow-up in `5f1907a9e0`: I kept the test in the existing sequential stage as requested, and completed the DYN-4061-specific work. The test now uses the lighter `agg.sh` path, sends one one-token request, and requires a debug receipt showing that the real canary reached the TRT-LLM handler with `priority=1.0`; the existing handler unit test also asserts that receipt and the `generate_async` priority.

Comment (knarangN) [reply tests/serve/test_trtllm.py]: Correction in `e733e818fb`: I removed the later follow-up changes. The final PR diff now only comments out `profiled_vram_gib(3.9)` and documents why this test must stay in the existing sequential GPU stage, exactly as suggested. No production logging, log-based assertion, launcher or request change, global guideline change, new marker, or scheduler change remains. The existing direct handler unit test continues to assert that the health-check payload reaches `generate_async` with `priority=1.0`.

**Review by nv-tusharma — APPROVED** [review]

**Comment by knarangN** [comment]
@ai-dynamo/dynamo-runtime-codeowners @ai-dynamo/dynamo-backend-trtllm-codeowners Could you please review this PR? The current head e733e818fb094e36e687e43c0981f2cf8dad7338 implements the reviewer-requested minimal change by keeping the startup-sensitive TRT-LLM health-check test in the existing sequential GPU stage. A current human approval is present; a qualifying code-owner approval is the remaining review gate. The failed PR jobs were rerun on the same SHA and the full PR workflow is now green: https://github.com/ai-dynamo/dynamo/actions/runs/32870096596. Exact-head H100 nightly validation is still running: https://github.com/ai-dynamo/dynamo/actions/runs/32870156966.

**Review by dagil-nvidia — APPROVED** [review]
Approving: all gates green at head e733e818 (CODEOWNERS drift 0 across 613 rules, strict coverage, 93 matcher tests, 26 skills validated), zero blocking findings, and the mechanism verified end to end - the sequential stage picks this test up in both nightly lanes that select it, no required marker category breaks, and unit coverage of the priority behavior survives the stage move. Note the process-codeowners request on this PR is vestigial: the final +3/-1 diff routes to runtime only, and the process- and trtllm-owned edits from earlier revisions were fully walked back (the human approval covers the exact final head).

Three non-blocking notes:

1. `tests/serve/test_trtllm.py:830` - `requested_trtllm_kv_tokens(2592)` is now inert on this test: it is consumed only through the parallel-orchestrator metadata path this test no longer enters, so the sequential run uses the engine's default KV allocation. Fold it under the same comment block, or add a clause saying it applies only if the profiled marker returns.
2. `tests/serve/test_trtllm.py:826` - the comment states the MPI startup stall as a general phenomenon while a sibling profiled TRT-LLM test stays in the parallel pool; one clause scoping it to this test's canary-at-startup pattern would stop it reading as an indictment of the whole pool.
3. PR metadata: the `documentation` label and the process/trtllm review-team requests are stale leftovers from withdrawn revisions.

Automated review pass; severity escalation stays with the human reviewers.

**Review by nv-tusharma — APPROVED** [review]

**Review by rmccorm4 — APPROVED** [review]
